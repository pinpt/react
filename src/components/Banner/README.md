### A simple notice banner

[Full Docs](https://react.preview.pinpoint.com/?path=/docs/components-banner) | [Github](https://github.com/pinpt/react/tree/master/src/components/Banner)

#### Theming

The parent className for styling this component is `.Pinpoint.Banner`

<details>
	<summary>Base Theme Styles</summary>

```css
.Pinpoint.Banner.wrapper {
	width: 100%;
	display: flex;
	align-items: center;
	justify-content: center;
	padding: 1rem;
	background-color: rgb(245, 158, 11);
	color: rgb(120, 53, 15);
}

.Pinpoint.Banner .inner {
	display: flex;
	align-items: center;
	flex-direction: row;
	width: 100%;
	max-width: 80rem;
}

.Pinpoint.Banner.icon {
	margin-right: 0.5rem;
	width: 1rem;
	height: 1rem;
}
```

</details>
