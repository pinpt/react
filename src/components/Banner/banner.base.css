.Pinpoint.Banner.wrapper {
	@apply w-full flex flex-row items-center justify-center p-4 bg-yellow-500 text-yellow-900;
}

.Pinpoint.Banner .close {
	@apply ml-auto cursor-pointer;
}

.Pinpoint.Banner .inner {
	@apply flex items-center flex-row w-full max-w-[75rem];
}

.Pinpoint.Banner .icon {
	@apply mr-2 w-4 h-4;
}
