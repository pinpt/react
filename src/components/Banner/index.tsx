import React from 'react';
import { faInfoCircle, faTimes, IconDefinition } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface BannerProps {
	icon?: IconDefinition;
	message?: React.ReactNode;
	children?: React.ReactNode;
	className?: string;
	closeable?: boolean;
	style?: React.CSSProperties;
}

const Banner = ({
	icon = faInfoCircle,
	message,
	children,
	className = 'bg-yellow-500 text-yellow-900',
	style,
	closeable,
}: BannerProps) => {
	const [visible, setVisible] = React.useState(true);
	if (!visible) {
		return null;
	}
	return (
		<div className={`Pinpoint Banner wrapper ${className}`} style={style}>
			{closeable && <FontAwesomeIcon icon={faTimes} className="close" onClick={() => setVisible(false)} />}
			<div className="inner">
				{message ? (
					<>
						{icon && <FontAwesomeIcon icon={icon} className="icon" />}
						<div>{message}</div>
					</>
				) : (
					<>{children}</>
				)}
			</div>
		</div>
	);
};

export default Banner;
