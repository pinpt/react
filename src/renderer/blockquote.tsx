import { registerNode, recurseIntoChildren, NodeProps } from './register';

const BlockQuote = ({ node }: NodeProps) => {
	return <blockquote>{recurseIntoChildren(node)}</blockquote>;
};

registerNode('blockquote', (node) => <BlockQuote node={node} />);
