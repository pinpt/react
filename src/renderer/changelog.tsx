import { ReactYouTubeLite as Youtube } from 'react-youtube-lite';
import { getFileUrl } from '../lib/file_metadata';
import { slugifyContent } from '../lib/string';
import { ChangelogCoverMedia, ChangelogMediaType } from '../lib/types';
import { Content as Doc } from './';

const ImageMedia = ({ src, title = '' }: { src: string; title?: string }) => {
	return (
		<div className="image">
			<img src={src} alt={title} />
		</div>
	);
};

const VideoMedia = ({ src }: { src: string }) => {
	return (
		<div className="video">
			<video controls src={src} />
		</div>
	);
};

const YoutubeMediaSSR = ({ id, poster = 'hqdefault' }: { id: string; poster?: string }) => {
	// the component isn't SSR friendly so this is basically used to sort of mimic it for generated code
	const posterUrl = `https://i.ytimg.com/vi/${id}/${poster}.jpg`;
	return (
		<>
			{poster && <link rel="preload" href={posterUrl} as="image" />}
			<div
				className="youtube ryt-lite embed-responsive aspect-ratio-16/9"
				style={{ backgroundImage: `url(${posterUrl})` }}
			>
				<div className="lty-playbtn" data-url={`https://www.youtube-nocookie.com/embed/${id}?autoplay=1`} />
			</div>
		</>
	);
};

const YoutubeMedia = ({
	id,
	metadata,
	staticMode,
}: {
	id: string;
	metadata?: Record<string, any>;
	staticMode?: true;
}) => {
	return (
		<div className="youtube">
			{!staticMode ? (
				<Youtube url={`https://www.youtube.com/watch?v=${id}`} poster={metadata?.poster} />
			) : (
				<YoutubeMediaSSR id={id} poster={metadata?.poster} />
			)}
		</div>
	);
};

const Media = ({ media, title, staticMode }: { media?: ChangelogCoverMedia; title?: string; staticMode?: true }) => {
	if (!media) {
		return <></>;
	}
	let content: React.ReactElement;
	switch (media.type) {
		case ChangelogMediaType.None: {
			return <></>;
		}
		case ChangelogMediaType.Image: {
			content = <ImageMedia src={getFileUrl(media.value)} title={title} />;
			break;
		}
		case ChangelogMediaType.Video: {
			content = <VideoMedia src={getFileUrl(media.value)} />;
			break;
		}
		case ChangelogMediaType.Youtube: {
			content = <YoutubeMedia id={media.value} metadata={media.metadata} staticMode={staticMode} />;
			break;
		}
	}
	return <div className="media-container">{content}</div>;
};

interface ChangelogProps {
	id: string;
	title?: string;
	content?: any;
	coverMedia?: ChangelogCoverMedia;
	limit?: number;
	divider?: boolean;
	staticMode?: true;
}

const ChangelogLog = (props: ChangelogProps) => {
	return (
		<article className="changelog notebook-editor read-only">
			{props.coverMedia && (
				<section className="covermedia">
					<Media media={props.coverMedia} title={props.title} staticMode={props.staticMode} />
				</section>
			)}
			<section className="ProseMirror">
				<Doc node={props.content} limit={props.limit} />
				{props.limit && props.content.content?.length > props.limit && (
					<div className="continue">
						<a href={slugifyContent(props.id, props.title)}>Continue Reading â†’</a>
					</div>
				)}
			</section>
			{props.divider && <hr className="divider" />}
		</article>
	);
};

export default ChangelogLog;
