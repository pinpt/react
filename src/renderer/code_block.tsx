import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { twilight } from 'react-syntax-highlighter/dist/cjs/styles/prism';
import { NodeProps, registerNode } from './register';

const CodeBlock = ({ node }: NodeProps) => {
	return (
		<SyntaxHighlighter
			language={node.attrs.language}
			style={twilight}
			customStyle={{
				background: 'black',
				fontFamily: 'ui-monospace',
				borderWidth: '0px',
				borderRadius: '.25em',
			}}
		>
			{node.content?.[0].text ?? ''}
		</SyntaxHighlighter>
	);
};

registerNode('code_block', (node) => <CodeBlock node={node} />);
