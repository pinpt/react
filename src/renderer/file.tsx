import {
	faDownload, faFile, faFileAlt, faFileArchive, faFileAudio, faFileCode, faFileExcel, faFilePdf,
	faFilePowerpoint, faFileVideo, faFileWord
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { NodeProps, registerNode } from './register';

export const iconForType = (type = '') => {
	if (/pdf/.test(type)) {
		return faFilePdf;
	}
	if (/zip|tar/.test(type)) {
		return faFileArchive;
	}
	if (/video/.test(type)) {
		return faFileVideo;
	}
	if (/audio/.test(type)) {
		return faFileAudio;
	}
	if (/excel|spreadsheet/.test(type)) {
		return faFileExcel;
	}
	if (/powerpoint/.test(type)) {
		return faFilePowerpoint;
	}
	if (/word/.test(type)) {
		return faFileWord;
	}
	if (/css|csv|graphql|html|java|json|php|python|typescript|xml/.test(type)) {
		return faFileCode;
	}
	if (/text/.test(type)) {
		return faFileAlt;
	}
	return faFile;
};

export const formattedSize = (size = 0) => {
	const kB = size / 1000;
	if (kB < 10) {
		return `${Math.round(kB * 10) / 10} KB`;
	}
	if (kB < 1000) {
		return `${Math.round(kB)} KB`;
	}
	const MB = kB / 1000;
	return `${Math.round(MB * 10) / 10} MB`;
};

const File = ({ node }: NodeProps) => {
	const { description, filename, publicUrl, size, type } = node.attrs;

	const onDownload = () => {
		if (publicUrl) {
			const u = new URL(publicUrl);
			u.searchParams.set('filename', filename);
			u.searchParams.set('disposition', 'attachment');
			const a = document.createElement('a');
			a.style.display = 'none';
			a.href = u.toString();
			// the filename you want
			a.download = filename;
			document.body.appendChild(a);
			a.click();
			a.remove();
		}
	};

	return (
		<div className="file">
			<div className="container">
				<FontAwesomeIcon icon={iconForType(type)} />
				<div className="details">
					{publicUrl ? (
						<a href={publicUrl} target="_blank" rel="noopener noreferrer" className="filename">
							{filename}
						</a>
					) : (
						<div className="filename">{filename}</div>
					)}
					{description && <div className="description">{description}</div>}
					<div className="size">{formattedSize(size)}</div>
				</div>
				<div className="options">
					{publicUrl && (
						<div className="option group" onClick={onDownload}>
							<FontAwesomeIcon icon={faDownload} />
							<span>Download</span>
						</div>
					)}
				</div>
			</div>
		</div>
	);
};

registerNode('file', (node) => <File node={node} />);
