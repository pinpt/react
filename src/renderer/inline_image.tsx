import { extractImageMetadataFromFileID } from '../lib/file_metadata';
import { NodeProps, registerNode } from './register';

const InlineImage = ({ node }: NodeProps) => {
	const { size } = extractImageMetadataFromFileID(node.attrs?.src);
	const width = size?.width;
	const height = size?.height;

	const image = (
		<img src={node.attrs?.src} width={width} height={height} alt={node.attrs?.alt} className="inline_image" />
	);

	return (
		<div>
			{node.attrs?.href ? (
				<a href={node.attrs.href} target="_blank" rel="noopener noreferrer">
					{image}
				</a>
			) : (
				<>{image}</>
			)}
		</div>
	);
};

registerNode('inline-image', (node) => <InlineImage node={node} />);
