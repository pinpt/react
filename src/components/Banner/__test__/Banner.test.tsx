import renderer from 'react-test-renderer';
import { faExclamationCircle } from '@fortawesome/free-solid-svg-icons';
import Banner from '../';

test('Test empty state', () => {
	const component = renderer.create(<Banner />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test with message', () => {
	const component = renderer.create(<Banner message="Hello world" />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test with message and icon', () => {
	const component = renderer.create(<Banner message="Hello world" icon={faExclamationCircle} />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test with children', () => {
	const component = renderer.create(<Banner>Hello World</Banner>);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test with className', () => {
	const component = renderer.create(<Banner className="bg-red-500 text-white">Hello World</Banner>);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test with closeable', () => {
	const component = renderer.create(<Banner closeable>Hello World</Banner>);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test with styles', () => {
	const component = renderer.create(<Banner style={{ backgroundColor: '#900', color: '#fff' }}>Hello World</Banner>);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
