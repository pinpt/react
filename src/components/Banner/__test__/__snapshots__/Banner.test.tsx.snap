// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test empty state 1`] = `
<div
  className="Pinpoint Banner wrapper bg-yellow-500 text-yellow-900"
>
  <div
    className="inner"
  />
</div>
`;

exports[`Test with children 1`] = `
<div
  className="Pinpoint Banner wrapper bg-yellow-500 text-yellow-900"
>
  <div
    className="inner"
  >
    Hello World
  </div>
</div>
`;

exports[`Test with className 1`] = `
<div
  className="Pinpoint Banner wrapper bg-red-500 text-white"
>
  <div
    className="inner"
  >
    Hello World
  </div>
</div>
`;

exports[`Test with closeable 1`] = `
<div
  className="Pinpoint Banner wrapper bg-yellow-500 text-yellow-900"
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-times fa-w-11 close"
    data-icon="times"
    data-prefix="fas"
    focusable="false"
    onClick={[Function]}
    role="img"
    style={Object {}}
    viewBox="0 0 352 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <div
    className="inner"
  >
    Hello World
  </div>
</div>
`;

exports[`Test with message 1`] = `
<div
  className="Pinpoint Banner wrapper bg-yellow-500 text-yellow-900"
>
  <div
    className="inner"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-info-circle fa-w-16 icon"
      data-icon="info-circle"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <div>
      Hello world
    </div>
  </div>
</div>
`;

exports[`Test with message and icon 1`] = `
<div
  className="Pinpoint Banner wrapper bg-yellow-500 text-yellow-900"
>
  <div
    className="inner"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-exclamation-circle fa-w-16 icon"
      data-icon="exclamation-circle"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <div>
      Hello world
    </div>
  </div>
</div>
`;

exports[`Test with styles 1`] = `
<div
  className="Pinpoint Banner wrapper bg-yellow-500 text-yellow-900"
  style={
    Object {
      "backgroundColor": "#900",
      "color": "#fff",
    }
  }
>
  <div
    className="inner"
  >
    Hello World
  </div>
</div>
`;
